src/server/terrain/chunkgen/StalagmiteGenerator.zig

```diff
@@ -0,0 +1,155 @@
+const std = @import("std");
+
+const main = @import("root");
+const random = main.random;
+const JsonElement = main.JsonElement;
+const terrain = main.server.terrain;
+const CaveMap = terrain.CaveMap;
+const CaveBiomeMap = terrain.CaveBiomeMap;
+const vec = main.vec;
+const Vec3d = vec.Vec3d;
+const Vec3f = vec.Vec3f;
+const Vec3i = vec.Vec3i;
+
+pub const id = "cubyz:stalagmite";
+
+pub const priority = 65537;
+
+pub const generatorSeed = 0x2ba5bef20d6153a5;
+
+var stalagmite: u16 = 0;
+
+const surfaceDist = 2; // How far away stalagmite can spawn from the wall.
+
+pub fn init(parameters: JsonElement) void {
+	_ = parameters;
+	// Find all the glow stalagmite ores:
+	// inline for(stalagmiteColor[0..], glowStalagmites[0..]) |color, *block| {
+	// 	const oreID = "cubyz:glow_stalagmite/" ++ color;
+	// 	block.* = main.blocks.getByID(oreID);
+	// }
+	stalagmite = main.blocks.getByID("cubyz:limestone");
+}
+
+pub fn deinit() void {
+
+}
+
+pub fn generate(worldSeed: u64, chunk: *main.chunk.ServerChunk, caveMap: CaveMap.CaveMapView, biomeMap: CaveBiomeMap.CaveBiomeMapView) void {
+	if(chunk.super.pos.voxelSize > 2) return;
+	const size = chunk.super.width;
+	// Generate caves from all nearby chunks:
+	var x = chunk.super.pos.wx -% main.chunk.chunkSize;
+	while(x != chunk.super.pos.wx +% size +% main.chunk.chunkSize) : (x +%= main.chunk.chunkSize) {
+		var y = chunk.super.pos.wy -% main.chunk.chunkSize;
+		while(y != chunk.super.pos.wy +% size +% main.chunk.chunkSize) : (y +%= main.chunk.chunkSize) {
+			var z = chunk.super.pos.wz -% main.chunk.chunkSize;
+			while(z != chunk.super.pos.wz +% size +% main.chunk.chunkSize) : (z +%= main.chunk.chunkSize) {
+				var seed = random.initSeed3D(worldSeed, .{x, y, z});
+				considerCoordinates(x, y, z, chunk, caveMap, biomeMap, &seed);
+			}
+		}
+	}
+}
+
+fn distSqr(x: f32, y: f32, z: f32) f32 {
+	return x*x + y*y + z*z;
+}
+
+fn considerStalagmite(x: i32, y: i32, z: i32, chunk: *main.chunk.ServerChunk, seed: *u64) void {
+	const relX: f32 = @floatFromInt(x -% chunk.super.pos.wx);
+	const relY: f32 = @floatFromInt(y -% chunk.super.pos.wy);
+	const relZ: f32 = @floatFromInt(z -% chunk.super.pos.wz);
+	
+	// Make some stalagmite spikes in random directions:
+	var spikes: f32 = 4;
+	spikes += random.nextFloat(seed)*spikes; // Use somewhat between spikes and 2*spikes spikes.
+	var _spike: f32 = 0;
+	while(_spike < spikes) : (_spike += 1) {
+		var length = 12 + random.nextFloat(seed) * random.nextFloat(seed) * 36;
+		const tiny = random.nextFloat(seed) < 0.9;
+		if (tiny) {
+			length = 2 * random.nextFloat(seed)*3;
+		}
+		// Choose a random direction:
+		const delX: f32 = 0;
+		const delY: f32 = 0;
```

Comment:
Please remove these always 0 constants and simplify the code that depends on them.